### DISCUSSION TITLE
How to show Android Fragment in Fabric Native Component ?  Couldn't find related documentation

### DISCUSSION THREAD

**Discussion by ndu-aspira** [reactwg/react-native-new-architecture#239] [Q&A] [answered]
Hi， 
I‘m trying to migrate my project to the new architecture.   Our map component was placed in an Android Fragment and it works well on the old architecture.   I tried to put the Fragment in the new Frbric Native Component following this guide https://github.com/reactwg/react-native-new-architecture/blob/main/docs/fabric-native-components.md.   It gives no error but displays blank screen. I put some log in my Map fragment.  The lifecycle method of the Fragment seems good ( on create, onCreateView, onStart..) all got invoked.

I also tried the simplest Fragment, ( a TextView in LinearLayout) and result is the same.

Is there any document of How to integrate Android Fragment into Fabric native component like the old one (https://reactnative.dev/docs/0.74/native-components-android#integration-with-an-android-fragment-example)

My rn version is 0.74.5
Any help is Appreciated.

My current code snippet. 
js side:

![image](https://github.com/user-attachments/assets/9a07c129-5a23-4fe4-af17-dfc1b84078c7)

![image](https://github.com/user-attachments/assets/2b7450c7-91cb-439e-8714-c158c1f76b65)

Android Native Side:
![image](https://github.com/user-attachments/assets/ad03d5d6-1516-4e9c-a951-76814275abf1)
![image](https://github.com/user-attachments/assets/247ebee1-3f0f-444a-a03a-6976d9d465ac)
![image](https://github.com/user-attachments/assets/142567eb-9bf0-4278-9aca-fd1a73933656)

Thanks.

**Comment by cortinico** [comment]
I've recently updated the integrating with Android Fragment guide here:
https://reactnative.dev/docs/next/integration-with-android-fragment

@ndu-aspira 
Could you make sure you follow it and report back if things are not working properly?

**Reply by ndu-aspira** [reply]
Hi @cortinico,
Thanks for your reply,  but I'm not looking for the guide of how to integrate RN into existing Full Screen Activity/Fragment.
Instead, I'm looking for a way to show Android Fragment in Fabric Native component like on the old architecture documented here：
https://reactnative.dev/docs/0.74/native-components-android.  

<img width="1043" alt="image" src="https://github.com/user-attachments/assets/18aae4c6-b913-47b3-9d7f-35ae22ce6ff8">

The corresponding new documentation (https://reactnative.dev/docs/next/legacy/native-components-android) is marked as unreleased and there's "legacy" in the doc path.  So I'm wondering if this part of doc is not updated for Fabric?

**Reply by cortinico** [reply]
> So I'm wondering if this part of doc is not updated for Fabric?

Yup we haven't updated that part of the doc to the New Architecture just yet.
Thanks for spotting it.

**Reply by ndu-aspira** [reply]
To be clear,  you mean the new architecture already support this feature only that the documentation is not updated right?  When can I expect the doc to be updated, or is there any sample/demo code snippet I can refer to get started for now...

Thanks very much.

**Reply by cortinico** [reply]
> To be clear, you mean the new architecture already support this feature only that the documentation is not updated right? 

Correct.

There are no sample/demos that I'm aware of sadly

**Reply by ndu-aspira** [reply]
Thank you anyway for the info.

**Reply by ndu-aspira** [reply]
@cortinico  just found the exact same issue here( https://github.com/facebook/react-native/issues/36432 ).  I looked at the PR you mentioned in that issue （https://github.com/facebook/react-native/pull/36263）, but it seems that it might not actually fix the problem described in the issue, right?

**Reply by cortinico** [reply]
The linked is issue is different from the one you're reporting.

**Reply by Vadko** [reply]
+1, any plans for documenting it?

**Reply by kevindice** [reply]
I'm also seeking to put a native fragment (from a vendor) into a React Native app as a Fabric component.  I've had a working implementation for the last year prior to enabling the New Architecture, but it renders as blank now.

**Reply by ndu-aspira** [reply, marked as answer]
After a lot of trying, I managed to make it work.  There're 2 part of changes need te be made in my case.

- do not use  ViewManagerDelegate, stick with receiveCommand.
- when manuallyLayoutChildren you need to specify the right and bottom your self.

attaching part of my view manager for reference.

```
ReactModule(name = GISMapViewManager.REACT_CLASS)
public class GISMapViewManager extends ViewGroupManager<FrameLayout> implements ReactContextProvider, CampsiteGisMapManagerInterface<FrameLayout> {

    private static final String TAG = "GISMapViewManager";

    public static final String REACT_CLASS = "CampsiteGisMap";
    public final int COMMAND_CREATE = 110;

    private final ReactApplicationContext reactContext;

    private final ViewManagerDelegate<FrameLayout> mDelegate;

    public GISMapViewManager(ReactApplicationContext reactContext) {
        this.reactContext = reactContext;
        MainApplication.reactApplicationContext = reactContext;
        mDelegate = new CampsiteGisMapManagerDelegate<>(this);
    }

//     @Nullable
//     @Override
//     protected ViewManagerDelegate<FrameLayout> getDelegate() {
//         return mDelegate;
//     }
 

    /**
     * Return a FrameLayout which will later hold the Fragment
     */
    @Override
    public FrameLayout createViewInstance(ThemedReactContext reactContext) {
        FrameLayout layout = new FrameLayout(reactContext);
        layout.setId(View.generateViewId());
        return layout;
    }

    /**
     * Map the "create" command to an integer
     */
    @Nullable
    @Override
    public Map<String, Integer> getCommandsMap() {
        return MapBuilder.of("create", COMMAND_CREATE);
    }

    /**
     * Handle "create" command (called from JS) and call createFragment method
     */
    @Override
    public void receiveCommand(@NonNull FrameLayout root, String commandId, @Nullable ReadableArray args) {
        super.receiveCommand(root, commandId, args);
        int reactNativeViewId = args.getInt(0);
        // int commandIdInt = Integer.parseInt(commandId);

        switch (commandId) {
            case "create":
                createFragment(root, reactNativeViewId);
                break;
            default: {
            }
        }
    }


    /**
     * Replace your React Native view with a custom fragment
     */
    public void createFragment(FrameLayout root, int reactNativeViewId) {
//        ViewGroup parentView = (ViewGroup) root.findViewById(reactNativeViewId);
//        setupLayout(parentView);
        setupLayout(root);

        SiteMapFragment siteMapFragment = SiteMapFragment.newInstance();
        siteMapFragment.setReactApplicationContext(reactContext);

        FragmentActivity activity = (FragmentActivity) reactContext.getCurrentActivity();
        activity.getSupportFragmentManager()
                .beginTransaction()
                .replace(reactNativeViewId, siteMapFragment, String.valueOf(reactNativeViewId))
                .commitNow()

    }

    public void setupLayout(View view) {
        Choreographer.getInstance().postFrameCallback(new Choreographer.FrameCallback() {
            @Override
            public void doFrame(long frameTimeNanos) {
                manuallyLayoutChildren(view);
                view.getViewTreeObserver().dispatchOnGlobalLayout();
                Choreographer.getInstance().postFrameCallback(this);
            }
        });
    }

    /**
     * Layout all children properly
     */
    public void manuallyLayoutChildren(View view) {
        // propWidth and propHeight coming from react-native props
        int width = propWidth;
        int height = propHeight;

        view.measure(
                View.MeasureSpec.makeMeasureSpec(width, View.MeasureSpec.EXACTLY),
                View.MeasureSpec.makeMeasureSpec(height, View.MeasureSpec.EXACTLY));

        view.layout(view.getLeft(),view.getTop(), view.getLeft() + width, view.getTop() + height);
    }

    @Override
    public ReactApplicationContext provideReactContext() {
        return reactContext;
    }

}
```

**Reply by kevindice** [reply]
Thank you, this helped a lot.
